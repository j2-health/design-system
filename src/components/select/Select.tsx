import { Select as AntDSelect, Spin } from 'antd'
import { LoadingOutlined } from '@ant-design/icons'
import type { SelectProps as AntDSelectProps } from 'antd'
import { CaretDown, MagnifyingGlass, XCircle } from '@phosphor-icons/react'
import { useState } from 'react'

type SelectProps = Expand<AntDSelectProps> & {
  size?: 'large' | 'middle' | 'small'
  mode?: '' | 'multiple'
  onChange?: (value: string | string[]) => void
  onSearch?: (value: string) => void
  options?: { label: string; value: string }[]
}

export const Select = (props: SelectProps) => {
  const [isFocused, setIsFocused] = useState(false)

  const handleSearch = (value: string) => {
    if (props.onSearch) {
      props.onSearch(value)
    }
  }

  return (
    <AntDSelect
      {...props}
      suffixIcon={
        props.loading ? (
          <Spin
            indicator={<LoadingOutlined spin />}
            size="small"
            data-testid="loading-spinner"
          />
        ) : props.showSearch && isFocused ? (
          <MagnifyingGlass
            size={14}
            data-testid="magnifying-glass"
            weight="duotone"
          />
        ) : (
          <CaretDown size={14} data-testid="caret-down" weight="duotone" />
        )
      }
      allowClear={
        props.allowClear
          ? { clearIcon: <XCircle size={14} weight="duotone" /> }
          : false
      }
      showSearch={props.showSearch}
      onSearch={props.onSearch ? handleSearch : undefined}
      onFocus={() => {
        setIsFocused(true)
      }}
      onBlur={() => {
        setIsFocused(false)
      }}
    />
  )
}
