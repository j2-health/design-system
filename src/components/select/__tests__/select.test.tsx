import { render, fireEvent } from '@testing-library/react'
import { Select } from '../Select'

describe('Select', () => {
  it('should render correctly', () => {
    const { container } = render(<Select />)
    expect(container).toMatchSnapshot()
  })

  it('should render with options', () => {
    const { container } = render(
      <Select
        options={[
          { label: 'Bronze Plan', value: 'bronze' },
          { label: 'Silver Plan', value: 'silver' },
          { label: 'Gold Plan', value: 'gold' },
          { label: 'Platinum Plan', value: 'platinum' },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should render with loading indicator', () => {
    const { container } = render(
      <Select
        loading
        options={[
          { label: 'Bronze Plan', value: 'bronze' },
          { label: 'Silver Plan', value: 'silver' },
          { label: 'Gold Plan', value: 'gold' },
          { label: 'Platinum Plan', value: 'platinum' },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should render with search enabled', () => {
    const { container } = render(
      <Select
        showSearch
        options={[
          { label: 'Bronze Plan', value: 'bronze' },
          { label: 'Silver Plan', value: 'silver' },
          { label: 'Gold Plan', value: 'gold' },
          { label: 'Platinum Plan', value: 'platinum' },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should call onSearch when search input changes', () => {
    const handleSearch = jest.fn()
    const { getByRole } = render(
      <Select
        showSearch
        onSearch={handleSearch}
        options={[
          { label: 'UHC', value: 'united_healthcare' },
          { label: 'Aetna', value: 'aetna' },
        ]}
      />
    )
    const input = getByRole('combobox')
    fireEvent.focus(input)
    fireEvent.change(input, { target: { value: 'UHC' } })
    expect(handleSearch).toHaveBeenCalledWith('UHC')
  })

  it('should render with allowClear enabled', () => {
    const { container } = render(
      <Select
        allowClear
        options={[
          { label: 'Bronze Plan', value: 'bronze' },
          { label: 'Silver Plan', value: 'silver' },
          { label: 'Gold Plan', value: 'gold' },
          { label: 'Platinum Plan', value: 'platinum' },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should not render clear icon when allowClear is disabled', () => {
    const { container } = render(
      <Select
        options={[
          { label: 'Bronze Plan', value: 'bronze' },
          { label: 'Silver Plan', value: 'silver' },
          { label: 'Gold Plan', value: 'gold' },
          { label: 'Platinum Plan', value: 'platinum' },
        ]}
      />
    )
    fireEvent.mouseOver(container.firstChild as Element)
    expect(container.querySelector('[data-testid="clear-icon"]')).toBeFalsy()
    expect(container).toMatchSnapshot()
  })
})
