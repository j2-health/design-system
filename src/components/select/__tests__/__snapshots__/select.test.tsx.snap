// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render correctly 1`] = `
<div>
  <div
    class="ant-select ant-select-outlined css-dev-only-do-not-override-ccdg5a ant-select-single ant-select-show-arrow"
  >
    <div
      class="ant-select-selector"
    >
      <span
        class="ant-select-selection-search"
      >
        <input
          aria-autocomplete="list"
          aria-controls="rc_select_TEST_OR_SSR_list"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="rc_select_TEST_OR_SSR_list"
          autocomplete="off"
          class="ant-select-selection-search-input"
          id="rc_select_TEST_OR_SSR"
          readonly=""
          role="combobox"
          style="opacity: 0;"
          type="search"
          unselectable="on"
          value=""
        />
      </span>
      <span
        class="ant-select-selection-placeholder"
      />
    </div>
    <span
      aria-hidden="true"
      class="ant-select-arrow"
      style="user-select: none;"
      unselectable="on"
    >
      <svg
        data-testid="caret-down"
        fill="currentColor"
        height="14"
        viewBox="0 0 256 256"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M208,96l-80,80L48,96Z"
          opacity="0.2"
        />
        <path
          d="M215.39,92.94A8,8,0,0,0,208,88H48a8,8,0,0,0-5.66,13.66l80,80a8,8,0,0,0,11.32,0l80-80A8,8,0,0,0,215.39,92.94ZM128,164.69,67.31,104H188.69Z"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`Select should render with loading indicator 1`] = `
<div>
  <div
    class="ant-select ant-select-outlined css-dev-only-do-not-override-ccdg5a ant-select-single ant-select-show-arrow ant-select-loading"
  >
    <div
      class="ant-select-selector"
    >
      <span
        class="ant-select-selection-search"
      >
        <input
          aria-autocomplete="list"
          aria-controls="rc_select_TEST_OR_SSR_list"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="rc_select_TEST_OR_SSR_list"
          autocomplete="off"
          class="ant-select-selection-search-input"
          id="rc_select_TEST_OR_SSR"
          readonly=""
          role="combobox"
          style="opacity: 0;"
          type="search"
          unselectable="on"
          value=""
        />
      </span>
      <span
        class="ant-select-selection-placeholder"
      />
    </div>
    <span
      aria-hidden="true"
      class="ant-select-arrow ant-select-arrow-loading"
      style="user-select: none;"
      unselectable="on"
    >
      <div
        aria-busy="true"
        aria-live="polite"
        class="ant-spin ant-spin-sm ant-spin-spinning css-dev-only-do-not-override-ccdg5a"
        data-testid="loading-spinner"
      >
        <span
          aria-label="loading"
          class="anticon anticon-loading anticon-spin ant-spin-dot"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="loading"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
            />
          </svg>
        </span>
      </div>
    </span>
  </div>
</div>
`;

exports[`Select should render with options 1`] = `
<div>
  <div
    class="ant-select ant-select-outlined css-dev-only-do-not-override-ccdg5a ant-select-single ant-select-show-arrow"
  >
    <div
      class="ant-select-selector"
    >
      <span
        class="ant-select-selection-search"
      >
        <input
          aria-autocomplete="list"
          aria-controls="rc_select_TEST_OR_SSR_list"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="rc_select_TEST_OR_SSR_list"
          autocomplete="off"
          class="ant-select-selection-search-input"
          id="rc_select_TEST_OR_SSR"
          readonly=""
          role="combobox"
          style="opacity: 0;"
          type="search"
          unselectable="on"
          value=""
        />
      </span>
      <span
        class="ant-select-selection-placeholder"
      />
    </div>
    <span
      aria-hidden="true"
      class="ant-select-arrow"
      style="user-select: none;"
      unselectable="on"
    >
      <svg
        data-testid="caret-down"
        fill="currentColor"
        height="14"
        viewBox="0 0 256 256"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M208,96l-80,80L48,96Z"
          opacity="0.2"
        />
        <path
          d="M215.39,92.94A8,8,0,0,0,208,88H48a8,8,0,0,0-5.66,13.66l80,80a8,8,0,0,0,11.32,0l80-80A8,8,0,0,0,215.39,92.94ZM128,164.69,67.31,104H188.69Z"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`Select should render with search enabled 1`] = `
<div>
  <div
    class="ant-select ant-select-outlined css-dev-only-do-not-override-ccdg5a ant-select-single ant-select-show-arrow ant-select-show-search"
  >
    <div
      class="ant-select-selector"
    >
      <span
        class="ant-select-selection-search"
      >
        <input
          aria-autocomplete="list"
          aria-controls="rc_select_TEST_OR_SSR_list"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="rc_select_TEST_OR_SSR_list"
          autocomplete="off"
          class="ant-select-selection-search-input"
          id="rc_select_TEST_OR_SSR"
          role="combobox"
          type="search"
          value=""
        />
      </span>
      <span
        class="ant-select-selection-placeholder"
      />
    </div>
    <span
      aria-hidden="true"
      class="ant-select-arrow"
      style="user-select: none;"
      unselectable="on"
    >
      <svg
        data-testid="caret-down"
        fill="currentColor"
        height="14"
        viewBox="0 0 256 256"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M208,96l-80,80L48,96Z"
          opacity="0.2"
        />
        <path
          d="M215.39,92.94A8,8,0,0,0,208,88H48a8,8,0,0,0-5.66,13.66l80,80a8,8,0,0,0,11.32,0l80-80A8,8,0,0,0,215.39,92.94ZM128,164.69,67.31,104H188.69Z"
        />
      </svg>
    </span>
  </div>
</div>
`;
